### PR TITLE
Decouple user credentials for external data sources from datasets

### PR DESCRIPTION
**Pull request by lebinh**
For PE-1167

This add a new API to DDFManager namely createDDF and a set of new abstraction over data source to load to Spark in io.ddf.ds package.

The old DataSource classes in io.ddf.datasource will be eventually deprecated and removed.

/cc @Huandao0812 @PangZhi @nhanitvn @mbbui @sameerp

### CODE REVIEW

**Comment by PangZhi** [comment]
@lebinh  Is there any test to show the usage of the new feature?

File: core/src/main/java/io/ddf/ds/User.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+   * Set the current user in the system.
+   * The current user is keep on a per-thread basis.
+   *
+   * @param user current user
+   */
+  public static void setCurrentUser(User user) {
+    currentUser.set(user);
+  }
+
+  private final String id;
```
Comment (darma-m) [inline core/src/main/java/io/ddf/ds/User.java]: Why not force id to be of type UUID?

**Comment by lebinh** [comment]
@PangZhi just added some tests, I am adding more.

File: core/src/main/java/io/ddf/ds/User.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  @Override
+  public String toString() {
+    return "User{" +
+        "id='" + id + '\'' +
+        ", username='" + username + '\'' +
+        '}';
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Comment (hai-adatao) [inline core/src/main/java/io/ddf/ds/User.java]: should have unit test for this

File: spark/src/test/scala/io/ddf/spark/ds/CreateDDFSuite.scala [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      | Location string
+    """.stripMargin.replaceAll("\n", "")
+
+  val sleep_schema = "StartTime int, SleepMinutes int, UID string, DeepSleepMinutes int, EndTime int"
+
+  override protected def beforeAll(): Unit = {
+    val user = new User("foo")
+    User.setCurrentUser(user)
+
+    val access_key = sys.env.get("S3_ACCESS")
```
Comment (hai-adatao) [inline spark/src/test/scala/io/ddf/spark/ds/CreateDDFSuite.scala]: Abstract this please, I saw another PR that use AWS_ACCESS_KEY

File: spark/src/main/java/io/ddf/spark/DelegatingDDFManager.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
+package io.ddf.spark;
```
Comment (nhanitvn) [inline spark/src/main/java/io/ddf/spark/DelegatingDDFManager.java]: Is it true that the purpose of this DelegatingDDFManager is for usage in test-cases only?

Comment (lebinh) [inline spark/src/main/java/io/ddf/spark/DelegatingDDFManager.java]: No, it's used in PE for CreateDDF API as well, as the design doc specify that there a one and only one mapping between data source <==> DDFManager, thus a singleton SparkDDFManager cannot be used for different source directly. Hence the need for this DelegatingDDFManager class.

File: spark/src/test/scala/io/ddf/spark/ds/CreateDDFSuite.scala (hunk 3 above)
Comment (lebinh) [inline spark/src/test/scala/io/ddf/spark/ds/CreateDDFSuite.scala]: I am following the name used in pyclient tests here

File: core/src/main/java/io/ddf/ds/User.java (hunk 2 above)
Comment (lebinh) [inline core/src/main/java/io/ddf/ds/User.java]: @hai-adatao you mean for this equals method? It is pretty standard and was auto generated actually.

Comment (hai-adatao) [inline core/src/main/java/io/ddf/ds/User.java]: I though you wrote it, if that's the case then it should be ok

File: spark/src/test/scala/io/ddf/spark/ds/CreateDDFSuite.scala [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    val username = getConfig(usernameKey)
+    val password = getConfig(passwordKey)
+    if (username.isDefined && password.isDefined)
+      Some(new UsernamePasswordCredential(username.get, password.get))
+    else
+      None
+  }
+
+  override protected def beforeEach(): Unit = {
+    val user = new User("foo")
```
Comment (nhanitvn) [inline spark/src/test/scala/io/ddf/spark/ds/CreateDDFSuite.scala]: As a single DDF user, should I care about User concept? @darma-m @lebinh

**Comment by lebinh** [comment]
retest this please

**Comment by lebinh** [comment]
@Huandao0812 @nhanitvn @PangZhi please review

**Comment by hai-adatao** [comment]
@lebinh please help rebase master first

**Comment by lebinh** [comment]
okay @hai-adatao I rebased to adatao branch

File: spark/src/test/scala/io/ddf/spark/ATestSuite.scala [hunk 6]
```diff
@@ -41,6 +38,41 @@ abstract class ATestSuite extends FunSuite with BeforeAndAfterEach with BeforeAn
...
+    *
+    * @param key
+    * @return
+    */
+  protected def getConfig(key: String): Option[String] = {
+    val sysEnvName = key.toUpperCase().replace(".", "_")
+    sys.props.get(key).orElse(sys.env.get(sysEnvName))
+  }
+
+  /**
```
Comment (nhanitvn) [inline spark/src/test/scala/io/ddf/spark/ATestSuite.scala]: Kool!!!

**Pull request merged**